File: apps/backend/src/services/standardize.service.ts
/**
 * @file apps/backend/src/services/standardize.service.ts
 * @description Data normalization and standardization service.
 * Converts raw invoice data from various document formats into a
 * consistent, normalized structure for the matching engine.
 *
 * Phase 1: Scaffold with placeholder implementation.
 * Phase 4: Implement actual normalization logic (string cleaning, date parsing, etc.)
 */

import type { RawInvoiceData } from '@gst/shared';
import type { IInvoice } from '../models/invoice.model.js';
import { derivePeriodFromDateStr, parseGstDate } from '../utils/date.utils.js';

// ---------------------------------------------------------------------------
// Normalization helpers
// ---------------------------------------------------------------------------

/** Common prefixes to strip from vendor names (with optional trailing dot/space) */
const VENDOR_PREFIX_RE =
  /^(?:m\/s\.?\s*|messrs\.?\s*)/i;

/** Suffix normalisation map (order matters: more specific first) */
const SUFFIX_REPLACEMENTS: [RegExp, string][] = [
  [/\bprivate\s+limited\b/gi, 'pvt ltd'],
  [/\bpvt\.?\s+ltd\.?\b/gi, 'pvt ltd'],
  [/\bprivate\s+ltd\.?\b/gi, 'pvt ltd'],
  [/\blimited\b/gi, 'ltd'],
  [/\bltd\.?\b/gi, 'ltd'],
];

/**
 * Normalizes a vendor name for consistent comparison during matching.
 *
 * Processing steps:
 *   1. Strip common Indian business prefixes (M/S, M/s., MESSRS)
 *   2. Normalise legal suffixes (PRIVATE LIMITED → pvt ltd, etc.)
 *   3. Replace `&` with `and`
 *   4. Remove remaining punctuation (keep letters, digits, spaces)
 *   5. Collapse multiple spaces, lowercase, trim
 *
 * Examples:
 *   `"M/S LIBERTY GLASS CRETIONS"` → `"liberty glass cretions"`
 *   `"SHARDA DOORS & PLYWOOD"`     → `"sharda doors and plywood"`
 *   `"M/S ABC TRADERS PVT. LTD."` → `"abc traders pvt ltd"`
 *
 * @param vendorName - Raw vendor name from source document
 * @returns Normalized vendor name string
 */
export function normalizeVendorName(vendorName: string): string {
  let name = vendorName.trim();

  // 1. Remove common prefixes
  name = name.replace(VENDOR_PREFIX_RE, '');

  // 2. Normalise legal suffixes
  for (const [pattern, replacement] of SUFFIX_REPLACEMENTS) {
    name = name.replace(pattern, replacement);
  }

  // 3. Replace & with and
  name = name.replace(/&/g, 'and');

  // 4. Remove punctuation except word chars and spaces
  name = name.replace(/[^a-zA-Z0-9\s]/g, ' ');

  // 5. Collapse spaces, lowercase, trim
  return name.replace(/\s+/g, ' ').trim().toLowerCase();
}

/**
 * Normalizes an invoice number for consistent comparison.
 *
 * Processing steps:
 *   1. Remove all non-alphanumeric characters
 *   2. Convert to uppercase
 *   3. Strip leading zeros only if the result is entirely numeric
 *      (preserves leading zeros in alphanumeric numbers like "INV001")
 *
 * Examples:
 *   `"INV/2024/001"` → `"INV2024001"`
 *   `"544"`          → `"544"`
 *   `"001"`          → `"1"`
 *   `"INV001"`       → `"INV001"`
 *
 * @param invoiceNumber - Raw invoice number from source document
 * @returns Normalized invoice number string
 */
export function normalizeInvoiceNumber(invoiceNumber: string): string {
  const stripped = invoiceNumber.toUpperCase().replace(/[^A-Z0-9]/g, '');

  // Strip leading zeros only for purely numeric invoice numbers
  if (/^\d+$/.test(stripped)) {
    return String(parseInt(stripped, 10));
  }

  return stripped;
}

/**
 * Derives the tax period (`YYYY-MM`) from an invoice date string.
 * Delegates to the shared `derivePeriodFromDateStr()` utility so ALL
 * six GST date formats are handled correctly.
 *
 * Examples:
 *   `"15-Mar-23"`   → `"2023-03"`
 *   `"25/03/2023"`  → `"2023-03"`
 *   `"2023-03-25"`  → `"2023-03"`
 *   `"15-03-23"`    → `"2023-03"`
 *   `"15.03.23"`    → `"2023-03"`
 *
 * @param invoiceDate - Invoice date string in any supported GST format
 * @returns Period string in `YYYY-MM` format
 */
export function derivePeriodFromDate(invoiceDate: string): string {
  return derivePeriodFromDateStr(invoiceDate);
}

// ---------------------------------------------------------------------------
// Return type compatible with IInvoice (minus Mongoose Document fields)
// ---------------------------------------------------------------------------

type StandardizedInvoice = Omit<IInvoice, keyof import('mongoose').Document | 'createdAt'>;

/**
 * Standardizes a batch of raw invoice records into objects ready for MongoDB insertion.
 *
 * For each raw invoice:
 *   - Normalizes vendor name and invoice number
 *   - Parses the invoice date into a JS Date (any supported GST format)
 *   - Derives the YYYY-MM period
 *   - Computes totalAmount with 2-decimal rounding
 *   - Sets default reconciliation fields (UNMATCHED, confidence 0, ITC null)
 *
 * @param rawInvoices - Array of raw invoice data objects
 * @param userId      - ID of the user the invoices belong to
 * @param source      - Source type: 'BOOKS' | 'GSTR_2A' | 'GSTR_2B'
 * @returns Array of standardized invoice objects ready for `InvoiceModel.insertMany()`
 */
export function standardizeInvoices(
  rawInvoices: RawInvoiceData[],
  userId: string,
  source: 'BOOKS' | 'GSTR_2A' | 'GSTR_2B',
): StandardizedInvoice[] {
  return rawInvoices.map((rawInvoice) => ({
    userId,
    source,
    gstin: rawInvoice.gstin,
    vendorName: rawInvoice.vendorName,
    normalizedVendorName: normalizeVendorName(rawInvoice.vendorName),
    invoiceNumber: rawInvoice.invoiceNumber,
    normalizedInvoiceNumber: normalizeInvoiceNumber(rawInvoice.invoiceNumber),
    invoiceDate: parseGstDate(rawInvoice.invoiceDate),
    period: derivePeriodFromDate(rawInvoice.invoiceDate),
    taxableAmount: rawInvoice.taxableAmount,
    igst: rawInvoice.igst,
    cgst: rawInvoice.cgst,
    sgst: rawInvoice.sgst,
    totalAmount:
      Math.round(
        (rawInvoice.taxableAmount + rawInvoice.igst + rawInvoice.cgst + rawInvoice.sgst) * 100,
      ) / 100,
    matchStatus: 'UNMATCHED' as const,
    itcCategory: null,
    matchConfidence: 0,
    description: rawInvoice.description,
  }));
}

File: apps/backend/app/services/standardize_service.py
"""
Standardization service — normalizes vendor names, invoice numbers, and derives periods.

Handles real-world GST data discrepancies between GSTR-2A and GSTR-2B:
  - "LIBERTY GLASS CRETIONS" (2A) vs "LIBERTY GLASS CREATIONS" (2B)
  - "SHARDA DOORS & PLYWOOD"  (2A) vs "SHARDHA DOORS AND PLYWOOD" (2B)
  - "PSI-25/26/02103"          (2B) vs "02103"                     (2A vch_no)
  - "INV/2024/001"             (2B) vs "INV2024001"                 (2A)
"""

import re
from app.models.invoice import Invoice
from app.utils.date_helpers import parse_gst_date, to_period


# ── Legal suffix normalizations ───────────────────────────────────────────────
LEGAL_SUFFIX_MAP = {
    "PRIVATE LIMITED": "PVT LTD",
    "PVT LIMITED":     "PVT LTD",
    "PRIVATE LTD":     "PVT LTD",
    "PVT. LTD.":       "PVT LTD",
    "PVT.LTD.":        "PVT LTD",
    "PVT. LIMITED":    "PVT LTD",
    "LIMITED":         "LTD",
    "LTD.":            "LTD",
    "CORPORATION":     "CORP",
    "ENTERPRISES":     "ENT",
}

# Prefixes to strip from vendor names
VENDOR_PREFIXES = [
    "M/S.", "M/S", "MESSRS.", "MESSRS",
    "MR.", "MR", "MS.", "MS",
    "SRI", "SHRI", "SMT",
]

# Characters to treat as word separators when normalizing invoice numbers
# We keep alphanumeric and then collapse separators so that:
#   "PSI-25/26/02103" → "PSI252602103"  (all separators removed)
#   "INV/2024/001"    → "INV2024001"
#   "02103"           → "2103"          (leading-zero strip)
_INV_SEPARATOR_RE = re.compile(r'[^A-Z0-9]+')
_LEADING_ZERO_RE  = re.compile(r'^0+(?=[A-Z])')   # strip leading zeros before a letter
_PURELY_NUMERIC   = re.compile(r'^\d+$')


def normalize_vendor_name(name: str) -> str:
    """
    Normalize vendor name for fuzzy matching.

    Steps
    -----
    1. Uppercase + strip whitespace.
    2. Remove common prefixes (M/S, MESSRS, SRI, etc.).
    3. Replace & / ' and ' with AND.
    4. Normalize legal suffixes (PRIVATE LIMITED → PVT LTD).
    5. Remove all punctuation except spaces.
    6. Collapse multiple spaces.

    The result is used only for matching (stored in normalized_vendor_name);
    the original vendor_name is preserved for display.
    """
    if not name:
        return ""

    result = name.upper().strip()

    # Strip prefixes
    for prefix in VENDOR_PREFIXES:
        if result.startswith(prefix + " "):
            result = result[len(prefix):].strip()
            break

    # Normalize & → AND, ' and ' → AND
    result = re.sub(r'\s+AND\s+', ' AND ', result.replace("&", " AND "))

    # Normalize legal suffixes (longest match first to avoid partial replacements)
    for long_form in sorted(LEGAL_SUFFIX_MAP, key=len, reverse=True):
        if result.endswith(" " + long_form) or result == long_form:
            result = result[: len(result) - len(long_form)].rstrip() + " " + LEGAL_SUFFIX_MAP[long_form]
            break

    # Remove all punctuation (keep only A-Z, 0-9, space)
    result = re.sub(r'[^A-Z0-9\s]', '', result)

    # Collapse spaces
    result = re.sub(r'\s+', ' ', result).strip()

    return result


def normalize_invoice_number(num: str) -> str:
    """
    Normalize invoice number for matching.

    Steps
    -----
    1. Strip + uppercase.
    2. Remove ALL separators (-, /, \\, ., spaces) — collapse to alphanumeric only.
       Rationale: "PSI-25/26/02103" and "PSI252602103" and "02103" should all
       produce a string that shares the critical numeric suffix "252602103".
       RapidFuzz token_set_ratio then handles the partial overlap.
    3. Strip leading zeros from purely numeric strings only:
         "002103" → "2103"
         "INV001" → "INV001"  (not stripped — has letters)
    4. Return "0" for an all-zero input (edge case).

    Note: the original invoice_number is preserved for display purposes.
    """
    if not num:
        return ""

    result = str(num).strip().upper()

    # Remove all non-alphanumeric characters (separators, slashes, dashes, etc.)
    result = _INV_SEPARATOR_RE.sub('', result)

    if not result:
        return ""

    # Strip leading zeros only for purely numeric invoice numbers
    if _PURELY_NUMERIC.match(result):
        stripped = result.lstrip('0')
        return stripped if stripped else '0'

    # For alphanumeric (e.g. "INV001"), strip only leading zeros before the
    # first alphabetic character (handles "00INV" → "INV" edge case)
    result = _LEADING_ZERO_RE.sub('', result)
    return result


def derive_period(date_str: str) -> str:
    """Parse any GST date format and return YYYY-MM period string."""
    if not date_str:
        return ""
    try:
        dt = parse_gst_date(str(date_str).strip())
        return to_period(dt)
    except (ValueError, TypeError):
        return ""


def batch_standardize(invoices: list[Invoice]) -> list[Invoice]:
    """
    Apply all normalizations to a list of Invoice documents in-place.

    Operations
    ----------
    - normalized_vendor_name    ← normalize_vendor_name(vendor_name)
    - normalized_invoice_number ← normalize_invoice_number(invoice_number)
    - period                    ← derived from invoice_date if not already set
    - total_amount              ← recomputed as sum of tax components (2 d.p.)
    """
    for invoice in invoices:
        invoice.normalized_vendor_name    = normalize_vendor_name(invoice.vendor_name)
        invoice.normalized_invoice_number = normalize_invoice_number(invoice.invoice_number)

        if not invoice.period:
            invoice.period = derive_period(str(invoice.invoice_date))

        # Recompute total_amount from components to fix any upstream rounding
        invoice.total_amount = round(
            (invoice.taxable_amount or 0.0)
            + (invoice.igst          or 0.0)
            + (invoice.cgst          or 0.0)
            + (invoice.sgst          or 0.0)
            + (invoice.cess          or 0.0),
            2,
        )

    return invoices